### PR TITLE
[feat] init create page

### PR DESCRIPTION
**Pull request by lucas-debize**

### CODE REVIEW

File: frontend/pages/create-project.tsx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  const [projectName, setProjectName] = useState('');
+  const [aboutProject, setAboutProject] = useState('');
+  const [selectedFile, setSelectedFile] = useState<File | null>(null);
+  const [material, setMaterial] = useState<string[]>([]);
+  const [materialsList, setMaterialList] = useState<string[]>([]);
+  const [newMaterial, setNewMaterial] = useState('');
+  const [listOfParticipant, setLists] = useState<string[]>([]);
+  const [newList, setNewList] = useState('');
+  const [error, setError] = useState<string>('');
+  const maxCharacterLimit = 1000;
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Do not use all os these `useState`, take a look a packages such as `formik` made to handle forms

File: frontend/pages/create-project.tsx [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      aboutProject,
+      listOfParticipant,
+      material,
+      selectedFile,
+    };
+    // To be removed (for test)
+    console.log('Data retrieved:', formData);
+    console.log('Data retrieved:', listOfParticipant);
+    console.log('Data retrieved:', material);
+    console.log('Selected File:', selectedFile);
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Should be deleted

File: frontend/pages/create-project.tsx [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+function CreateProject() {
+  const [projectName, setProjectName] = useState('');
+  const [aboutProject, setAboutProject] = useState('');
+  const [selectedFile, setSelectedFile] = useState<File | null>(null);
+  const [material, setMaterial] = useState<string[]>([]);
+  const [materialsList, setMaterialList] = useState<string[]>([]);
+  const [newMaterial, setNewMaterial] = useState('');
+  const [listOfParticipant, setLists] = useState<string[]>([]);
+  const [newList, setNewList] = useState('');
+  const [error, setError] = useState<string>('');
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: You never use `setError`, is this legacy ?

File: frontend/pages/create-project.tsx [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    if (newList.trim() !== '') {
+      setLists([...listOfParticipant, newList]);
+      setNewList('');
+    }
+  };
+  const removeParticipant = (index: number) => {
+    const updatedGuests = [...listOfParticipant];
+    updatedGuests.splice(index, 1);
+    setLists(updatedGuests);
+  };
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: You won't need this when you'll be using a proper package to handle forms

File: frontend/pages/project.tsx [hunk 5]
```diff
@@ -9,6 +10,9 @@ const Project: React.FC<{ projects: string[] }> = ({ projects }) => {
           <li key={index}>{project}</li>
         ))}
       </ul>
+      <Link href="/create-project">
+        Create Project
+      </Link>
```
Comment (thmarinho) [inline frontend/pages/project.tsx]: Aren't you missing some css here ?

**Review by thmarinho — CHANGES_REQUESTED** [review]

File: frontend/pages/create-project.tsx [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            // Handle form submission
+        },
+    });
+
+    const handleAboutProjectChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+        const inputValue = e.target.value;
+        if (inputValue.length <= maxCharacterLimit) {
+            formik.setFieldValue('aboutProject', inputValue);
+        }
+    };
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Use the `maxlength` attribute instead of this function

File: frontend/pages/create-project.tsx [hunk 7]
```diff
@@ -0,0 +1,221 @@
+import React, {useState, useEffect} from 'react';
+import Link from 'next/link';
+import { PhotoIcon } from '@heroicons/react/24/solid';
+import { useFormik } from 'formik';
+
+interface CreateProjectProps {}
+
+const CreateProject: React.FC<CreateProjectProps> = () => {
+    const maxCharacterLimit = 20;
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Do declare this as a variable, just send it as a props to the textarea
Make this number bigger

File: frontend/pages/create-project.tsx [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        const inputValue = e.target.value;
+        if (inputValue.length <= maxCharacterLimit) {
+            formik.setFieldValue('aboutProject', inputValue);
+        }
+    };
+
+    const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        const file = e.currentTarget.files?.[0];
+        formik.setFieldValue('selectedFile', file);
+    };
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Don't send the whole file, read the content as base64
A visual feedback would be nice on upload completion, maybe a preview

File: frontend/pages/create-project.tsx [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+    const addParticipant = () => {
+        if (formik.values.newList.trim() !== '') {
+            formik.setFieldValue('listOfParticipant', [...formik.values.listOfParticipant, formik.values.newList]);
+            formik.setFieldValue('newList', '');
+        }
+    };
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx:45]: I don't think you need this

File: frontend/pages/create-project.tsx [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+    const removeParticipant = (index: number) => {
+        formik.setFieldValue('listOfParticipant', formik.values.listOfParticipant.filter((_, i) => i !== index));
+    };
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Don't use index, look for a matching email

File: frontend/pages/create-project.tsx [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+    const addMaterial = () => {
+        if (formik.values.newMaterial.trim() !== '' && isMaterialValid(formik.values.newMaterial)) {
+            formik.setFieldValue('material', [...formik.values.material, formik.values.newMaterial]);
+            formik.setFieldValue('newMaterial', '');
+            formik.setFieldValue('error', '');
+        } else {
+            formik.setFieldValue('error', 'This material is not available');
+        }
+    };
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Autocompletion on elements could be nicer (you don't allow student to use non existing materials)
In that case, this code will become legacy

File: frontend/pages/create-project.tsx [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+                        <label htmlFor="cover-photo" className="block text-sm font-medium leading-6 text-gray-900">
+                            Cover photo
+                        </label>
+                        <div className="mt-2 flex justify-center rounded-lg border border-dashed border-gray-900/25 px-6 py-10">
+                            <div className="text-center">
+                                <PhotoIcon className="mx-auto h-12 w-12 text-gray-300" aria-hidden="true" />
+                                <div className="mt-4 flex text-sm leading-6 text-gray-600">
+                                    <label htmlFor="file-upload" className="relative cursor-pointer rounded-md bg-white font-semibold text-indigo-600 focus-within:outline-none focus-within:ring-2 focus-within:ring-indigo-600 focus-within:ring-offset-2 hover:text-indigo-500">
+                                        <span>Upload a file</span>
+                                        <input id="file-upload" type="file" className="sr-only" onChange={(e) => handleFileChange(e)} />
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx:196]: Allow multiple files and use the `accept` props, only images should be allowed

File: frontend/pages/create-project.tsx [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+                            </label>
+                            <div className="mt-2">
+                                <textarea
+                                    id="about"
+                                    name="aboutProject"
+                                    rows={4}
+                                    className="p-1 block w-full md:w-auto rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 md:text-sm md:leading-6"
+                                    value={formik.values.aboutProject}
+                                    onChange={handleAboutProjectChange}
+                                />
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Make the element bigger, this should content all of the project description

File: frontend/pages/create-project.tsx [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+                            </div>
+                            <p className="text-sm text-gray-500">{`${formik.values.aboutProject.length}/${maxCharacterLimit} characters`}</p>
+                        </div>
+
+                        <div className="md:col-span-3 md:ml-12 lg:ml-20">
+                            <label htmlFor="material" className="block text-sm font-medium leading-6 text-gray-900">
+                                Material
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: plural

File: frontend/pages/create-project.tsx [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+                                    type="text"
+                                    id="listOfParticipant"
+                                    name="newList"
+                                    className="mr-2 p-1 block w-full md:w-auto rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 md:text-sm md:leading-6"
+                                    value={formik.values.newList}
+                                    onChange={formik.handleChange}
+                                />
+                                <button type="button" className="bg-blue-500 text-white text-sm p-1 rounded-md" onClick={addParticipant}>
+                                    Add
+                                </button>
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx:107]: Pressing enter/tab key instead of pressing the button would be cool

**Review by thmarinho — CHANGES_REQUESTED** [review]

**Comment by thmarinho** [comment]
Please rebase your branch. This way, the CI will run its tests

File: frontend/pages/create-project.tsx (hunk 8 above)
Comment (lucas-debize) [reply frontend/pages/create-project.tsx]: i did a preview and read in base64

File: frontend/pages/create-project.tsx (hunk 15 above)
Comment (lucas-debize) [reply frontend/pages/create-project.tsx:107]: i only change when enter is pressed

File: frontend/package.json [hunk 16]
```diff
@@ -13,9 +13,11 @@
   "dependencies": {
     "@headlessui/react": "^1.7.17",
     "@heroicons/react": "^2.0.18",
+    "formik": "^2.4.5",
     "next": "14.0.2",
     "react": "^18",
-    "react-dom": "^18"
+    "react-dom": "^18",
+    "yup": "^1.3.2"
```
Comment (thmarinho) [inline frontend/package.json]: Why is this pacakge installed ? You're not using it

File: frontend/pages/create-project.tsx [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+import Link from 'next/link';
+import { CheckCircleIcon, PhotoIcon } from '@heroicons/react/24/solid';
+import { useFormik } from 'formik';
+import Image from 'next/image';
+
+interface CreateProjectProps {
+    maxCharacterLimit: number;
+}
+
+const CreateProject: React.FC<CreateProjectProps> = ({ maxCharacterLimit }) => {
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Why is `maxCharacterLimit` a prop ? It isn't send by another component

File: frontend/pages/create-project.tsx [hunk 18]
```diff
@@ -0,0 +1,225 @@
+import React from 'react';
+import Link from 'next/link';
+import { CheckCircleIcon, PhotoIcon } from '@heroicons/react/24/solid';
+import { useFormik } from 'formik';
+import Image from 'next/image';
+
+interface CreateProjectProps {
+    maxCharacterLimit: number;
+}
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]: Delete this

File: frontend/pages/create-project.tsx [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+
+const CreateProject: React.FC<CreateProjectProps> = ({ maxCharacterLimit }) => {
+    maxCharacterLimit = 500;
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx]:
```suggestion
    const projectDescriptionMaxLength = 1500;
```

File: frontend/pages/create-project.tsx [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+        action: () => void
+    ) => {
+        if (e.key === 'Enter') {
+            e.preventDefault();
+            action();
+        }
+    };
+
+    const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        const file = e.currentTarget.files?.[0];
```
Comment (thmarinho) [inline frontend/pages/create-project.tsx:36]:
```suggestion
        const file = e.currentTarget.files?.[0];
        const reader = new FileReader();
        reader.readAsDataURL(file as Blob);
        reader.onloader = (content) => console.log(content);
```

**Review by thmarinho — CHANGES_REQUESTED** [review]

**Review by thmarinho — APPROVED** [review]

**Pull request merged**